### ISSUE TITLE
[BUG] Failed to upgrade 4 nodes from v1.1.2 to v1.2.0-rc4, stuck in post-draining on the third node

### ISSUE THREAD

**Issue by TachunLin** [harvester/harvester#4285]
**Describe the bug**

* When upgrade 4 nodes Harvester from v1.1.2 to v1.2.0-rc4, the upgrade process stuck in `Post-draining` on the third node.

  ![image](https://github.com/harvester/harvester/assets/29251855/4c64f2b6-c4b9-4580-a0f1-eb34f6d2285c)

* Check the post-drain pod, contains `Failed to download the requested file from "http://upgrade-repo-hvst-upgrade-vzxdd.harvester-system/harvester-iso/rootfs.squashfs` 
  ```
  [[ 34 -lt 100 ]]                                                                                                                                         │
  │ + curl -sSfL http://upgrade-repo-hvst-upgrade-vzxdd.harvester-system/harvester-iso/rootfs.squashfs -o /host/usr/local/upgrade_tmp/tmp.RXRFJZN9TE           │
  │ curl: (7) Failed to connect to upgrade-repo-hvst-upgrade-vzxdd.harvester-system port 80 after 2 ms: Couldn't connect to server                             │
  │ Failed to download the requested file from "http://upgrade-repo-hvst-upgrade-vzxdd.harvester-system/harvester-iso/rootfs.squashfs" to "/host/usr/local/upg │
  │ + echo 'Failed to download the requested file from "http://upgrade-repo-hvst-upgrade-vzxdd.harvester-system/harvester-iso/rootfs.squashfs" to "/host/usr/l │
  │ + sleep 10 
  ```
* And finally stuck in `curl -sfL http://upgrade-repo-hvst-upgrade-vzxdd.harvester-system/harvester-iso/harvester-release.yaml `
  ```
  Wait for upgrade repo ready...                                                                                                                             │
  │ + curl -sfL http://upgrade-repo-hvst-upgrade-vzxdd.harvester-system/harvester-iso/harvester-release.yaml                                                   │
  │ + echo 'Wait for upgrade repo ready...'                                                                                                                    │
  │ + sleep 5            
  ```

* The third node keep in `cordened` state
  ![image](https://github.com/harvester/harvester/assets/29251855/8676a9f7-39e5-4e69-9efb-77bee30d6698)

* The `prometheus-rancher-monitoring-prometheus-0` pod use extremely high cpu and memory usage
  ![image](https://github.com/harvester/harvester/assets/29251855/f6c0d5a7-6ae6-446f-a066-a6de91348172)

* The `virt-launcher-upgrade-repo-hvst-upgrade-vzxdd-xlgjz` pod use extremely high cpu and memory usage
  ![image](https://github.com/harvester/harvester/assets/29251855/6044ff86-7efe-40f9-8c67-36ae8a6315ad)

**To Reproduce**
1. Prepare a four nodes v1.1.2 Harvester cluster 
2. Create an external vlan network
3. Create two vms, one set external vlan and another use management network 
4. Prepare offline  upgrade to v1.2.0-rc4 
5. Set the increase upgrade deadline 
6. Monitor the upgrade process 

**Expected behavior**

Should be able to complete the four nodes upgrade to v1.2.0-rc4 without hanging on the post-draining state of any node

**Support bundle**

[supportbundle_6377d75a-271c-433b-a5e9-0d440c87c922_2023-07-20T17-08-43Z.zip](https://github.com/harvester/harvester/files/12112279/supportbundle_6377d75a-271c-433b-a5e9-0d440c87c922_2023-07-20T17-08-43Z.zip)

**Environment**
 - Harvester ISO version: v1.1.2 to v1.2.0-rc4
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): 4 nodes bare machine

**Comment by bk201** [comment]
The upgrade VM's volume is dead:
```
harvester-tlg8g:~ # journalctl -u iscsid | grep 87f
Jul 20 13:43:46 harvester-tlg8g iscsid[6154]: iscsid: Connection4:0 to [target: iqn.2019-10.io.longhorn:pvc-87f1face-d896-43e7-9748-7ddfc06385bc, portal: 10.52.3.12,3260] through [iface: default] is operational now
Jul 20 16:07:06 harvester-tlg8g iscsid[6154]: iscsid: Connection4:0 to [target: iqn.2019-10.io.longhorn:pvc-87f1face-d896-43e7-9748-7ddfc06385bc, portal: 10.52.3.12,3260] through [iface: default] is shutdown.
Jul 20 16:07:31 harvester-tlg8g iscsid[6154]: iscsid: Connection6:0 to [target: iqn.2019-10.io.longhorn:pvc-87f1face-d896-43e7-9748-7ddfc06385bc, portal: 10.52.3.43,3260] through [iface: default] is operational now
```
Thus the post-drain script can't download the squashfs imge.
cc @Vicente-Cheng 

Also seeing I/O error on existing VMs:
```
│ default                     4m12s                         Warning                    IOerror                    virtualmachineinstance/vm1                    243322                      │
│ default                     119s                          Warning                    IOerror                    virtualmachineinstance/vm2                    250229                      │
```

**Comment by Vicente-Cheng** [comment]
After discussing with the Longhorn team member (@Shuo-wu), Thanks for the helping!

We found this issue would be like a split-brain issue.

We found the upgrade VM is not accessible because the root fs volume `pvc-87f1face-d896-43e7-9748-7ddfc06385bc` is temporarily down. That means the iSCSI was re-login, and the VM was stuck. So the post-drain (actually system upgrade) was stuck.

And why the `pvc-87f1face-d896-43e7-9748-7ddfc06385bc` is temporarily down? That is because the older im-e was clean up.

The whole timeline would be like follows:
1. Create an Upgrade VM
2. Upgrade Longhorn (at this moment, the old im-e do not be deleted due to the upgrade VM (pvc) depending on it)

3. Somehow, a split-brain-like situation happened on three nodes, `harvester-ftpn4`, `harvester-tlg8g` and `harvester-djm5h`. The split-brain-like situation would cause the owner to change between the above nodes (like the following logs). Also, that will lead the objects (instance/instance-manager) being updated to state `Unknown`.

The `harvester-ftpn4` logs:
```
2023-07-20T16:07:02.612781774Z stderr F time="2023-07-20T16:07:02Z" level=info msg="Engine got new owner harvester-ftpn4" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-ftpn4
2023-07-20T16:07:06.0170417Z stderr F time="2023-07-20T16:07:06Z" level=info msg="Engine got new owner harvester-ftpn4" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-ftpn4
2023-07-20T16:07:09.04804195Z stderr F time="2023-07-20T16:07:09Z" level=info msg="Engine got new owner harvester-ftpn4" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-ftpn4
2023-07-20T16:07:09.97288767Z stderr F time="2023-07-20T16:07:09Z" level=info msg="Engine got new owner harvester-ftpn4" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-ftpn4
2023-07-20T16:07:10.686847955Z stderr F time="2023-07-20T16:07:10Z" level=info msg="Engine got new owner harvester-ftpn4" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-ftpn4
2023-07-20T16:07:11.241817881Z stderr F time="2023-07-20T16:07:11Z" level=info msg="Engine got new owner harvester-ftpn4" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-ftpn4
```

The `harvester-tlg8g` logs:
```
2023-07-20T16:07:06.042557907Z stderr F time="2023-07-20T16:07:06Z" level=info msg="Engine got new owner harvester-tlg8g" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-tlg8g
2023-07-20T16:07:06.56078795Z stderr F time="2023-07-20T16:07:06Z" level=info msg="Engine got new owner harvester-tlg8g" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-tlg8g
2023-07-20T16:07:08.362743097Z stderr F time="2023-07-20T16:07:08Z" level=info msg="Engine got new owner harvester-tlg8g" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-tlg8g
2023-07-20T16:07:09.135252725Z stderr F time="2023-07-20T16:07:09Z" level=info msg="Engine got new owner harvester-tlg8g" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-tlg8g
2023-07-20T16:07:09.650335703Z stderr F time="2023-07-20T16:07:09Z" level=info msg="Engine got new owner harvester-tlg8g" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-tlg8g
2023-07-20T16:07:09.926205857Z stderr F time="2023-07-20T16:07:09Z" level=info msg="Engine got new owner harvester-tlg8g" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-tlg8g
2023-07-20T16:07:10.249680818Z stderr F time="2023-07-20T16:07:10Z" level=info msg="Engine got new owner harvester-tlg8g" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-tlg8g
2023-07-20T16:07:10.626545025Z stderr F time="2023-07-20T16:07:10Z" level=info msg="Engine got new owner harvester-tlg8g" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-tlg8g
2023-07-20T16:07:10.958066199Z stderr F time="2023-07-20T16:07:10Z" level=info msg="Engine got new owner harvester-tlg8g" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-tlg8g
2023-07-20T16:07:11.164628903Z stderr F time="2023-07-20T16:07:11Z" level=info msg="Engine got new owner harvester-tlg8g" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-tlg8g
2023-07-20T16:07:11.491254995Z stderr F time="2023-07-20T16:07:11Z" level=info msg="Engine got new owner harvester-tlg8g" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-tlg8g
2023-07-20T16:07:11.916414549Z stderr F time="2023-07-20T16:07:11Z" level=info msg="Engine got new owner harvester-tlg8g" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-tlg8g
2023-07-20T16:07:12.239285685Z stderr F time="2023-07-20T16:07:12Z" level=info msg="Engine got new owner harvester-tlg8g" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-tlg8g
2023-07-20T16:07:12.635769988Z stderr F time="2023-07-20T16:07:12Z" level=info msg="Engine got new owner harvester-tlg8g" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-tlg8g
2023-07-20T16:07:13.510299059Z stderr F time="2023-07-20T16:07:13Z" level=info msg="Engine got new owner harvester-tlg8g" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-tlg8g
2023-07-20T16:07:14.053592517Z stderr F time="2023-07-20T16:07:14Z" level=info msg="Engine got new owner harvester-tlg8g" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-tlg8g
2023-07-20T16:07:14.602653667Z stderr F time="2023-07-20T16:07:14Z" level=info msg="Engine got new owner harvester-tlg8g" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-tlg8g
2023-07-20T16:07:15.041876424Z stderr F time="2023-07-20T16:07:15Z" level=info msg="Engine got new owner harvester-tlg8g" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-tlg8g
```
 
The `harvester-djm5h` logs:
```
2023-07-20T16:07:10.677502166Z stderr F time="2023-07-20T16:07:07Z" level=info msg="Engine got new owner harvester-djm5h" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-djm5h
2023-07-20T16:07:10.677656382Z stderr F time="2023-07-20T16:07:08Z" level=info msg="Engine got new owner harvester-djm5h" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-djm5h
2023-07-20T16:07:11.506355712Z stderr F time="2023-07-20T16:07:11Z" level=info msg="Engine got new owner harvester-djm5h" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-djm5h
2023-07-20T16:07:11.894705389Z stderr F time="2023-07-20T16:07:11Z" level=info msg="Engine got new owner harvester-djm5h" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-djm5h
2023-07-20T16:07:12.396253982Z stderr F time="2023-07-20T16:07:12Z" level=info msg="Engine got new owner harvester-djm5h" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-djm5h
2023-07-20T16:07:12.70088135Z stderr F time="2023-07-20T16:07:12Z" level=info msg="Engine got new owner harvester-djm5h" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-djm5h
2023-07-20T16:07:13.56447389Z stderr F time="2023-07-20T16:07:13Z" level=info msg="Engine got new owner harvester-djm5h" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-djm5h
2023-07-20T16:07:14.401927158Z stderr F time="2023-07-20T16:07:14Z" level=info msg="Engine got new owner harvester-djm5h" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-djm5h
2023-07-20T16:07:14.656317176Z stderr F time="2023-07-20T16:07:14Z" level=info msg="Engine got new owner harvester-djm5h" controller=longhorn-engine engine=pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef node=harvester-djm5h
```

4. We could see that on the time between 16:07:06 - 16:07:15.
The engine `pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef` can not decide which one is the owner.
The instance (engine) status was set to Unknown because the owner changed. As follows logs:
```
2023-07-20T16:07:02.612892249Z stderr F time="2023-07-20T16:07:02Z" level=warning msg="The related node harvester-tlg8g of instance pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef is down or deleted, will mark the instance as state UNKNOWN"
``` 

5. Because the instance (engine) status is Unknown, somehow the problematic node `harvester-tlg8g` still can fetch the old im-e then the behavior of `syncInstanceManager` would try to clean up the old im-e (refer to this code block https://github.com/longhorn/longhorn-manager/blob/432b6d40f02a6fdffd8389df9d779016775f1d3a/controller/node_controller.go#L970-L985).
So we could see the following logs:
```
2023-07-20T16:07:05.789228425Z stderr F time="2023-07-20T16:07:05Z" level=debug msg="Prepare to clean up the redundant instance manager instance-manager-e-6235db52 when there is no running/starting instance"
2023-07-20T16:07:05.952625055Z stderr F time="2023-07-20T16:07:05Z" level=debug msg="Prepare to clean up the redundant instance manager instance-manager-e-6235db52 when there is no running/starting instance"
2023-07-20T16:07:06.092388028Z stderr F time="2023-07-20T16:07:06Z" level=debug msg="Prepare to clean up the redundant instance manager instance-manager-e-6235db52 when there is no running/starting instance"
2023-07-20T16:07:08.781007664Z stderr F time="2023-07-20T16:07:08Z" level=debug msg="Prepare to clean up the redundant instance manager instance-manager-e-6235db52 when there is no running/starting instance"
2023-07-20T16:07:09.055078635Z stderr F time="2023-07-20T16:07:09Z" level=debug msg="Prepare to clean up the redundant instance manager instance-manager-e-6235db52 when there is no running/starting instance"
2023-07-20T16:07:09.619045295Z stderr F time="2023-07-20T16:07:09Z" level=debug msg="Prepare to clean up the redundant instance manager instance-manager-e-6235db52 when there is no running/starting instance"
2023-07-20T16:07:09.990984105Z stderr F time="2023-07-20T16:07:09Z" level=debug msg="Prepare to clean up the redundant instance manager instance-manager-e-6235db52 when there is no running/starting instance"
2023-07-20T16:07:10.330468226Z stderr F time="2023-07-20T16:07:10Z" level=debug msg="Prepare to clean up the redundant instance manager instance-manager-e-6235db52 when there is no running/starting instance"
2023-07-20T16:07:10.596063625Z stderr F time="2023-07-20T16:07:10Z" level=debug msg="Prepare to clean up the redundant instance manager instance-manager-e-6235db52 when there is no running/starting instance"
2023-07-20T16:07:11.032527981Z stderr F time="2023-07-20T16:07:11Z" level=debug msg="Prepare to clean up the redundant instance manager instance-manager-e-6235db52 when there is no running/starting instance"
2023-07-20T16:07:11.301060449Z stderr F time="2023-07-20T16:07:11Z" level=debug msg="Prepare to clean up the redundant instance manager instance-manager-e-6235db52 when there is no running/starting instance"
2023-07-20T16:07:11.543060041Z stderr F time="2023-07-20T16:07:11Z" level=debug msg="Prepare to clean up the redundant instance manager instance-manager-e-6235db52 when there is no running/starting instance"
2023-07-20T16:07:11.981588824Z stderr F time="2023-07-20T16:07:11Z" level=debug msg="Prepare to clean up the redundant instance manager instance-manager-e-6235db52 when there is no running/starting instance"
2023-07-20T16:07:12.701130568Z stderr F time="2023-07-20T16:07:12Z" level=debug msg="Prepare to clean up the redundant instance manager instance-manager-e-6235db52 when there is no running/starting instance"
2023-07-20T16:07:13.946641691Z stderr F time="2023-07-20T16:07:13Z" level=debug msg="Prepare to clean up the redundant instance manager instance-manager-e-6235db52 when there is no running/starting instance"
2023-07-20T16:07:14.342761649Z stderr F time="2023-07-20T16:07:14Z" level=debug msg="Prepare to clean up the redundant instance manager instance-manager-e-6235db52 when there is no running/starting instance"
2023-07-20T16:07:15.084413579Z stderr F time="2023-07-20T16:07:15Z" level=debug msg="Prepare to clean up the redundant instance manager instance-manager-e-6235db52 when there is no running/starting instance"
2023-07-20T16:07:16.01684496Z stderr F time="2023-07-20T16:07:16Z" level=debug msg="Prepare to clean up the redundant instance manager instance-manager-e-6235db52 when there is no running/starting instance"
```

6. Even finally the engine gets the right owner, the old im-e is already deleted (so it will try to find the new im-e but the engine is still owned by the old im-e). Then this would cause the engine to stop. Then the new im-e would spawn it. Like follows logs:
```
2023-07-20T16:07:24.450789503Z stderr F time="2023-07-20T16:07:24Z" level=warning msg="Cannot find the instance status in instance manager instance-manager-e-2857d9b4851b35016b89aad48bb6d220
for the running instance pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef, will mark the instance as state ERROR"
2023-07-20T16:07:24.450852941Z stderr F time="2023-07-20T16:07:24Z" level=debug msg="Instance handler updated instance pvc-87f1face-d896-43e7-9748-7ddfc06385bc-e-20afd9ef state, old state unk
nown, new state error"
```

These issues would be seen as two main problems.
1. Does the etcd/kube-api stable?

  I found some logs about applying too long and some warning timeout on the etcd log for the node`harvester-tlg8g`,
  but that would cause the instance-manager to be checked as `Down` or `NotReady`? I am unsure about it, maybe
  we could dig more into it.

2. Longhorn-manager could have been more tolerant of the split-brain-like situation.

  After the discussion, we could open a Longhorn GitHub issue for it.
  https://github.com/longhorn/longhorn/issues/6479

**Comment by bk201** [comment]
https://github.com/longhorn/longhorn/issues/6479 is fixed in Longhorn v1.6.0.
The issue is hard to reproduce, please re-open if see this again.
